### PR TITLE
AO3-6003 Add i18n-tasks to normalize locale files and add tests to detect missing and unused translation keys

### PR DESCRIPTION
**Pull request by Bilka2**
# Pull Request Checklist

* [X] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [X] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [X] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [X] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [X] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6003

## Purpose

Updated version of #4472 per the discussion there.

Install and configure the i18n-tasks gem.
Enable tests for missing, unused locale keys and normalized locale files (all English only).
Fix 3 broken locales: [`comments.check_permission_to_edit.error.frozen`](https://github.com/otwcode/otwarchive/commit/783f5ea1ce49597225d3f43565a89b3dce002f6b), [`blocked.users.index.heading.landmark.blocked_users` and `muted.users.index.heading.landmark.muted_users`](https://github.com/otwcode/otwarchive/commit/5cc16f6bd866c40a33d309a8f43b33db0c9df839).
Ignore incorrect reporting of unused locales/mark locales as used where possible.
Add test for ensuring no locale values contain newlines. (English only)
Remove newlines from [signup_notification mailer locale](https://github.com/otwcode/otwarchive/commit/ffaf2607f95feab847a659568d790064b2e9ceb4).
Ignore_missing for locale keys where the default value is defined in the inside `t()` call in the .rb file.
Normalize English locale files.

## Testing Instructions

I manually verified that the 3 locales I fixed are broken on production and fixed in dev. Likely these should also be verified during QA.

## Notes

The tests are configured to run only for the English locale. If they would run for all locales, "missing" would simply report missing translations ([1000](https://gist.github.com/Bilka2/c96920d14599186c8480c49f73334e7d#file-missing-txt)) and "normalized" expects different formatting than the phrase export produces, more info below. The tests that make sense to run for all locales are the "unused" test and the "newlines" test. However the problems would have to be fixed, probably in phrase?

### Test: Unused keys (all locales)
Reports 120 unused keys. Some of these are related to b636d94ec9f550efeb37ed8ae90da8f3f61dcd73, but most are legitimate unused keys, mostly in the mailers. Output from the task: https://gist.github.com/Bilka2/c96920d14599186c8480c49f73334e7d#file-unused-txt

### Test: Newlines in values (all locales)

The following locale keys have values that contain newlines:
`["fa.user_mailer.signup_notification.text.part3", "fr.user_mailer.signup_notification.text.part3", "id.user_mailer.invitation_to_claim.text.part2", "pt-BR.user_mailer.invitation_to_claim.text.part7", "uk.user_mailer.anonymous_or_unrevealed_notification.anonymous_info"]`

### Multi locale configuration problems

The gem does not allow routing locale writes by language. This means that [moving locales](https://otwarchive.atlassian.net/browse/AO3-6240) will place the locale in the wrong files. The config has a commented out line to make it work correctly for any locale besides en.

Moving locales is further complicated by the fact that the gem normalizes files that locale is moved to/in. The gem disagrees how files should be normalized compared to phrase: Alphabetical sorting is slightly different (`users` vs `user_mailer`) and handling of line_length (when configured) is different.

So [AO3-6240](https://otwarchive.atlassian.net/browse/AO3-6240) will not work as is, but may work with some workarounds:
* Uncommenting line 25 in the config
* Running the move tasks
* Manually merging the produced en.yml file with the mailers/en.yml file
* (optional) manually undoing the different normalization of the other locale files before the phrase import.
* Removing the line in the config again 

## Credit

Bilka (he/him)

[AO3-6240]: https://otwarchive.atlassian.net/browse/AO3-6240?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Bilka2** [comment]
Testing instructions for the 3 previously broken locales:
`comments.check_permission_to_edit.error.frozen`
* Log in and create a work/open a work you own.
* Leave a comment on your work.
* Click "Edit" on that comment.
* In a new tab, freeze the thread with that comment.
* In the original tab, finish editing the comment by clicking "Update".
* It should flash an error message. (I had to refresh the page and it was shown at the top of the work.)
* The error message should say "Frozen comments cannot be edited."
* It should not say "translation missing: en.comment.check_permission_to_edit.error.frozen"

`blocked.users.index.heading.landmark.blocked_users` and `muted.users.index.heading.landmark.muted_users`
* The affected URLs are https://archiveofourown.org/users/USERNAME/blocked/users and https://archiveofourown.org/users/USERNAME/muted/users
* Inspect the HTML, find a h3 element with the `landmark heading` classes after the "Block the user" / "Mute the user" form.
* The h3 element should not contain a span with the `translation_missing` class.
* The h3 element should not contain a stray " %>".
* The h3 element should contain the text "Listing Blocked Users" / "Listing Muted Users".

**Review by redsummernight — COMMENTED** [review]
Looks good!

> The tests that make sense to run for all locales are the "unused" test and the "newlines" test. However the problems would have to be fixed, probably in phrase?

Yep, that's fine. We'll talk to the Translation team to fix those in Phrase and export to GitHub, then we'll enable these tests for all locales.

> The config has a commented out line to make it work correctly for any locale besides en.

I think this is fine. We can deal with en manually because we split its locale files in a different way. If the move tool works for all the other locales, we're good.

> Moving locales is further complicated by the fact that the gem normalizes files that locale is moved to/in. The gem disagrees how files should be normalized compared to phrase.

This is also fine. Once we move the keys, we'll import the fixes to Phrase. Then the next time Phrase does an export, those files will be reformatted in the way Phrase prefers.

File: spec/lib/i18n/i18n_tasks_spec.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  it "en files are normalized" do
+    non_normalized = i18n.non_normalized_paths.select { |path| path.end_with?("en.yml") }
+    error_message = "The following files need to be normalized:\n" \
+                    "#{non_normalized.map { |path| "  #{path}" }.join("\n")}\n" \
+                    "Please run `i18n-tasks normalize -l en' to fix"
+    expect(non_normalized).to be_empty, error_message
+  end
+
+  it "en does not have values with newlines" do
```
Comment (redsummernight) [inline spec/lib/i18n/i18n_tasks_spec.rb:28]: Can you add a comment here, "We want to keep formatting out of I18n-ized strings, and one paragraph per key."?

File: config/i18n-tasks.yml [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+## Reporting locale, default: en. Available: en, ru.
+# internal_locale: en
+
+# Read and write translations.
+data:
+  ## Translations are read from the file system. Supported format: YAML, JSON.
+  ## Provide a custom adapter:
+  # adapter: I18n::Tasks::Data::FileSystem
+
+  # Locale files or `File.find` patterns where translations are read from:
```
[the remark this answers is not in the document]
Comment (Bilka2) [reply config/i18n-tasks.yml:16]: Sure looks like it, fixed it here and made a PR upstream: https://github.com/glebm/i18n-tasks/pull/507

File: app/views/user_mailer/creatorship_notification.html.erb [hunk 3]
```diff
@@ -1,6 +1,9 @@
 <% creation_type = @creation.class.name.underscore %>
 <% content_for :message do %>
   <p>
+    <%# i18n-tasks-use t('user_mailer.creatorship_notification.intro_chapter')
+      i18n-tasks-use t('user_mailer.creatorship_notification.intro_series')
+      i18n-tasks-use t('user_mailer.creatorship_notification.intro_work')-%>
```
[the remark this answers is not in the document]
Comment (Bilka2) [reply app/views/user_mailer/creatorship_notification.html.erb:6]: Thank you!

**Comment by Bilka2** [comment]
Great that the workaround will be enough for the move tasks. I'll grab the issue since it's a bit of an ordeal and I already figured out how to make it work.

If you think moving locales will be needed for more than that one issue I could document the task list for "how to move locale" on the i18n page on the wiki.

File: config/i18n-tasks.yml [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+#   google_translate_api_key: "AbC-dEf5"
+#   # DeepL Pro Translate
+#   # Get an API key and subscription at https://www.deepl.com/pro to use DeepL Pro
+#   deepl_api_key: "48E92789-57A3-466A-9959-1A1A1A1A1A1A"
+#   # deepl_host: "https://api.deepl.com"
+#   # deepl_version: "v2"
+
+## Do not consider these keys missing:
+ignore_missing:
+  - comments.check_permission_to_modify_{frozen,hidden}_status.permission_denied
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply config/i18n-tasks.yml:117]: You can use the full key, `action_name` will give you the name of the action: https://guides.rubyonrails.org/v6.1/action_controller_overview.html#routing-parameters.

**Comment by redsummernight** [comment]
> Great that the workaround will be enough for the move tasks. I'll grab the issue since it's a bit of an ordeal and I already figured out how to make it work.
> 
> If you think moving locales will be needed for more than that one issue I could document the steps on the i18n page on the wiki.

Thanks for taking that! I added a note on [AO3-6240](https://otwarchive.atlassian.net/browse/AO3-6240) to update the wiki afterward.

File: config/i18n-tasks.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bilka2) [reply config/i18n-tasks.yml]: Okay, it now uses the full key. These 4 translations are covered by the automated tests in `comments_controller_spec.rb`.

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -211,14 +211,22 @@ def check_permission_to_edit
   def check_permission_to_modify_frozen_status
     return if permission_to_modify_frozen_status
 
-    flash[:error] = t(".permission_denied")
```
Comment (redsummernight) [inline app/controllers/comments_controller.rb:214]: You could write `t("comments.#{action_name}.permission_denied")` and keep the `i18n-tasks-use` hints. Subjective but I'd prefer that over an if statement without else in this case.

**Review by redsummernight — APPROVED** [review]
Thank you!

**Pull request merged**